File: Programmers/[12979] 기지국 설치/ONE/12979.java
class Solution {
    public int solution(int n, int[] stations, int w) {
        int start = 1, boundary = 1 + w * 2, answer = 0;

        for (int station : stations) {
            int stationStart = station - w;
            if (stationStart > start)
                answer += countStation(stationStart - start, boundary);
            start = station + w + 1;
        }

        if (start <= n)
            answer += countStation(n - start + 1, boundary);

        return answer;
    }

    private int countStation(int n, int boundary) {
        int count = n / boundary;
        if (n % boundary != 0)
            count += 1;
        return count;
    }
}

File: Programmers/[12979] 기지국 설치/Johoseong/12979.py
def solution(n, stations, w):
    answer = 0
    not_5g = []

    s, e = 1, 1
    for station in stations:
        e = station - w
        not_5g.append(e - s)
        s = station + w + 1
    if (stations[-1] + w) < n:
        not_5g.append(n + 1 - s)

    for cnt in not_5g:
        if cnt <= 0: continue
        if cnt <= (2 * w + 1):
            answer += 1
        else:
            answer += (cnt // (2 * w + 1))
            if cnt % (2 * w + 1) != 0:
                answer += 1
    return answer

print(solution(16, [1, 6, 14], 1))